src/net_processing.cpp

```diff
@@ -1677,11 +1689,22 @@ void static ProcessGetData(CNode& pfrom, const CChainParams& chainparams, CConnm
             continue;
         }
 
-        CTransactionRef tx = FindTxForGetData(pfrom, inv.hash, mempool_req, longlived_mempool_time);
+        CTransactionRef tx = FindTxForGetData(pfrom, inv.hash, mempool_req, now);
         if (tx) {
             int nSendFlags = (inv.type == MSG_TX ? SERIALIZE_TRANSACTION_NO_WITNESS : 0);
             connman->PushMessage(&pfrom, msgMaker.Make(nSendFlags, NetMsgType::TX, *tx));
             mempool.RemoveUnbroadcastTx(inv.hash);
+            // As we're going to send tx, make sure its unconfirmed parents are made requestable.
+            for (const auto& txin : tx->vin) {
+                auto txinfo = mempool.info(txin.prevout.hash);
+                if (txinfo.tx && txinfo.m_time > now - UNCONDITIONAL_RELAY_DELAY) {
+                    // Relaying a transaction with a recent but unconfirmed parent.
+                    if (WITH_LOCK(pfrom.m_tx_relay->cs_tx_inventory, return !pfrom.m_tx_relay->filterInventoryKnown.contains(txin.prevout.hash))) {
+                        LOCK(cs_main);
+                        State(pfrom.GetId())->m_recently_announced_invs.insert(txin.prevout.hash);
```

Comment:
I'm not sure I quite follow the conversation here, so apologies if this is redundant. I was just looking at this commit and wondering if we might insert too many entries into the bloom filter -- being willing to add all unconfirmed parents of a transaction makes the bloom filter effectively 24x smaller than we'd expect it to be.

For outbound peers, I think we're only within a factor of 2 of being able to roll the filter assuming no unconfirmed parents are added to it, so this seems potentially problematic?  It doesn't seem to me like RBF is a necessary precondition, just a busy network with people relaying lots of transactions with lots of unconfirmed parents.

Reply:
So if you have a transaction A with (say) 2000 outputs, and a transaction B that spends all 2000 of those outputs, then I believe this logic would cause us to enter A's txid 2000 times into the filter, which I believe will waste thousands of filter entries, right?

Reply:
@sdaftuar Where does the number 24x and 2x come from? I think you may be right that there is a mismatch; `INVENTORY_MAX_RECENT_RELAY` is sized based on the avg _outbound_ inv rate, while the auto-adding of unconfirmed parents may mean that indirectly everything added to the mempool in the past `UNCONDITIONAL_RELAY_DELAY` can be announced - is there any reason why there'd be a lower bound than that?

You're also right about the duplicate txid issue, which seems solvable independently using a set to deduplicate or so.

Reply:
@sipa On further thought I think my concern about the size due to transaction chains may be incorrect -- I forgot that we add transactions that we announce to a peer to that peer's `filterInventoryKnown`, which means that such transactions wouldn't be re-added to the peer's recently announced filter.  So the only situation where you get this potential issue of adding a much larger multiple of the outbound peer announcement rate is where a node has just recently connected (and therefore never got the INVs for the parents), but that should solve itself relatively quickly (and wouldn't be a general issue on the network in busy times, as I initially hypothesized). 

> You're also right about the duplicate txid issue, which seems solvable independently using a set to deduplicate or so.

Yeah I think this gets fixed immediately by using the mempool to tell us the unconfirmed parents of the transaction in question, rather than loop over inputs.

(PS If you open a PR to fix this, could you please wait until after wtxidrelay is merged?  Many thanks :) )